### DISCUSSION TITLE
Parsing to/from a Tagged value

### DISCUSSION THREAD

**Discussion by oliverfoggin** [pointfreeco/swift-url-routing#69] [Q&A]
I'm trying to introduce the `Tagged` framework into our project and just doing a PoC now.

I have a struct like...

```
struct Item {
  let id: String
}
```

and a route...

```
enum ServerRoute {
  case items(ItemsRoute)

  enum ItemsRoute {
    case list
    case item(id: String, ItemRoute)

    enum ItemRoute {
      case detail
      case delete
    }
  }
}
```

Something like this anyway.

My Router at the moment is able to parse this well and embed the `id` from the `Item` into the url so you end up with a path like...

```
/items
/items/123abc
/items/123abc/delete
```

I'm putting the id into the path using a simple ParserPrinter like...

```
Path { Parse(.string) }
```

But now I'd like to update this to make Item like...

```
struct Item {
  let id: Id

  typealias Id = Tagged<Item, String>
}
```

I've got these changes into the app except now I'm not sure how to update the Router ParserPrinter to accept this `Item.Id` instead of just String.

I've tried like...

```
Path { Parse(.string).map { Item.Id($0) } }
```

But that didn't work. It told me `Initializer 'init(_:with:)' requires that <LONG NESTED PARSER NAME> conforms to ParserPrinter`.

I also tried using the `Parse(_ transform, with: )` init too with the same error (pretty much).

Please could someone advise on how to do this?

Thanks

**Comment by oliverfoggin** [comment]
Ooh... ok, possibly got a bit further now. I've managed to get this building...

```
Parse(.memberwise(Item.Id.init(rawValue:)), with: { Parse(.string) })
```

It feels like that's not as good as it could be though. 😄

**Reply by oliverfoggin** [reply]
And now I've found this... `Parse(.representing(Item.Id.self), with: { Parse(.string) })`

Which feels a bit better.

Is that the intended way?

Thanks

**Reply by oliverfoggin** [reply]
And now this... `Parse(.string).map(.representing(Item.Id.self))` which I feel is much better.

**Reply by stephencelis** [reply]
@oliverfoggin Conversions and parsers live in a weird overlapping world where you can substitute many for one another. We're definitely open to ideas how to make fewer possible permutations of the same thing, but we're also not sure if it's possible.

The most succinct version of the above would be:

```swift
Parse(.string.representing(Item.Id.self))
```

(Most `map`-pable methods on conversions are also available via dot-chaining.)
